node/src/store.rs

```diff
@@ -1,9 +1,136 @@
+use celestia_types::ExtendedHeader;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use tendermint::Hash;
+use thiserror::Error;
+use tracing::{error, info};
+
+use crate::exchange::ExtendedHeaderExt;
+
 #[derive(Debug)]
-pub struct Store {}
+pub struct Store {
+    headers: DashMap<Hash, ExtendedHeader>,
+    height_to_hash: DashMap<u64, Hash>,
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum InsertError {
+    // TODO: do we care about the distinction
+    #[error("Hash {0} already exists in store")]
+    HashExists(Hash),
+    #[error("Height {0} already exists in store")]
+    HeightExists(u64),
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum ReadError {
+    // TODO: should we roll internal errors into one
+    #[error("Store in inconsistent state, lost head")]
+    LostStoreHead,
+    #[error("Store in inconsistent state, height->hash mapping exists, but hash doesn't")]
+    LostHash,
+    #[error("Failed to convert height from usize to u64, should not happen")]
+    HeadHeightConversionError,
+
+    #[error("Header not found in store")]
+    NotFound,
+}
 
 impl Store {
     pub fn new() -> Self {
-        Store {}
+        Store {
+            headers: DashMap::new(),
+            height_to_hash: DashMap::new(),
+        }
+    }
+
+    pub fn get_head_height(&self) -> usize {
+        self.height_to_hash.len()
+    }
+
+    pub fn append_header(&self, header: ExtendedHeader) -> Result<(), InsertError> {
+        let hash = header.hash();
+        let height = header.height(); // TODO: should Store be responsible for making sure we have
+                                      // all 1..=head headers?
```

Comment:
It shouldn't. Ideally `Syncer` should be able to parallelize the filling of the `Store`, so it will never be able to have `1..=head` headers. What store should be responsible is returning all the continues ranges we have in the store and the missing ranges.

Reply:
Per our yesterday's discussion it seems we want Store to keep continuous range from genesis to current verified head and keep syncer responsible for catching up and feeding the store headers in ordered fasion. I'll change the store code and add tests.